src/algorithms/nolip.jl

```diff
@@ -0,0 +1,223 @@
+set_next_direction!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = state.d .= .-state.res
+set_next_direction!(iter::NOLIPIteration, state::NOLIPState) = set_next_direction!(acceleration_style(typeof(iter.directions)), iter, state)
+
+update_direction_state!(::QuasiNewtonStyle, ::NOLIPIteration, state::NOLIPState) = update!(state.H, state.x - state.x_prev, state.res - state.res_prev)
+update_direction_state!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = return
+update_direction_state!(iter::NOLIPIteration, state::NOLIPState) = update_direction_state!(acceleration_style(typeof(iter.directions)), iter, state)
+
+reset_direction_state!(::QuasiNewtonStyle, ::NOLIPIteration, state::NOLIPState) = reset!(state.H)
+reset_direction_state!(::NoAccelerationStyle, ::NOLIPIteration, state::NOLIPState) = return
+reset_direction_state!(iter::NOLIPIteration, state::NOLIPState) = reset_direction_state!(acceleration_style(typeof(iter.directions)), iter, state)
+
+function Base.iterate(iter::NOLIPIteration{R}, state::NOLIPState) where R
+    f_Az, a, b, c = R(Inf), R(Inf), R(Inf), R(Inf)
+
+    # store iterate and residual for metric update later on
+    state.x_prev .= state.x
+    state.res_prev .= state.res
+
+    # compute FBE
+    FBE_x = f_model(iter, state) + state.g_z
+
+    sigma = iter.beta * (0.5 / state.gamma) * (1 - iter.alpha)
+    tol = 10 * eps(R) * (1 + abs(FBE_x))
+    threshold = FBE_x - sigma * norm(state.res)^2 + tol
+
+    tau_backtracks = 0
+    can_update_direction = true
+
+    while true
+
+        if can_update_direction
+            # compute direction
+            set_next_direction!(iter, state)
+            # backtrack tau 1 → 0
+            state.tau = R(1)
+            state.x .= state.x_prev .+ state.d
+        else
+            state.x .= (1 - state.tau) * (state.x_prev .- state.res_prev) + state.tau * (state.x_prev .+ state.d)
+        end
+
+        mul!(state.Ax, iter.A, state.x)
+        state.f_Ax = gradient!(state.grad_f_Ax, iter.f, state.Ax)
+        mul!(state.At_grad_f_Ax, adjoint(iter.A), state.grad_f_Ax)
+
+        state.y .= state.x .- state.gamma .* state.At_grad_f_Ax
+        state.g_z = prox!(state.z, iter.g, state.y, state.gamma)
+        state.res .= state.x .- state.z
+
+        f_Az_upp = f_model(iter, state)
+
+        if (iter.gamma === nothing || iter.adaptive == true)
+            mul!(state.Az, iter.A, state.z)
+            f_Az = gradient!(state.grad_f_Az, iter.f, state.Az)
+            tol = 10 * eps(R) * (1 + abs(f_Az))
+            if f_Az > f_Az_upp + tol && state.gamma >= iter.minimum_gamma
+                state.gamma *= 0.5
+                if state.gamma < iter.minimum_gamma
+                    @warn "stepsize `gamma` became too small ($(state.gamma))"
+                end
+                can_update_direction = true
+                reset_direction_state!(iter, state)
+                continue
+            end
+        end
```

Comment:
I agree, it would be much better. However, `backtrack_stepsize` can potentially take several iterations before returning, whereas here we backtrack only once before restarting.